java-dynamic-sqs-listener-core/src/main/java/com/jashmore/sqs/broker/singlethread/SingleThreadedMessageBroker.java

```diff
@@ -0,0 +1,139 @@
+ */
+@Slf4j
+public class SingleThreadedMessageBroker extends AbstractMessageBroker {
+    private final MessageRetriever messageRetriever;
+    private final MessageProcessor messageProcessor;
+    private final ExecutorService executorService;
+
+    public SingleThreadedMessageBroker(final MessageRetriever messageRetriever, final MessageProcessor messageProcessor) {
+        this(messageRetriever, messageProcessor, Executors.newFixedThreadPool(2));
+    }
+
+    @VisibleForTesting
+    SingleThreadedMessageBroker(final MessageRetriever messageRetriever,
+                                final MessageProcessor messageProcessor,
+                                final ExecutorService executorService) {
+        super(executorService);
+
+        this.messageRetriever = messageRetriever;
+        this.messageProcessor = messageProcessor;
+        this.executorService = executorService;
+    }
+
+    @Override
+    protected SingleThreadMessageController createController(final CompletableFuture<Object> controllerFinishedFuture) {
+        return new SingleThreadMessageController(messageRetriever, messageProcessor, executorService, controllerFinishedFuture);
+    }
+
+    /**
+     * This is the controller that will handle the retrieval of messages and the subsequent processing.
+     *
+     * <p>This will only be run on a single thread and any message that is processed will be started on a new thread. The reason for starting it on a new
+     * thread is because we don't want interrupts to cause the processing of the message to be stopped. If the thread is interrupted it will wait until the
+     * message is processed for stopping the thread.
+     */
+    @Slf4j
+    @RequiredArgsConstructor
+    static class SingleThreadMessageController  implements AbstractMessageBroker.Controller {
+        /**
+         * The amount of time that the thread should be slept if there was an error retrieving messages for processing.
+         */
+        private static final int BACKOFF_TIME_MS = 1000;
+
+        private final MessageRetriever messageRetriever;
+        private final MessageProcessor messageProcessor;
+        private final ExecutorService executorService;
+        private final CompletableFuture<Object> controllerFinishedFuture;
+
+        /**
+         * When shutting down this will determine whether any messages currently being processed should be interrupted.
+         */
+        private boolean interruptThreads;
+
+        @Override
+        public void run() {
+            log.debug("Starting message listener");
+            Future<?> currentlyRunningThread = null;
+            boolean shouldStop = false;
+            while (!shouldStop) {
+                final Message message;
+                try {
+                    message = messageRetriever.retrieveMessage();
+
+                    currentlyRunningThread = executorService.submit(() -> messageProcessor.processMessage(message));
```

Comment:
And in this case, just grab the message inline, no need to put in another thread since its blocking and its the only thread. I would just get rid of this class altogether.

Reply:
Yes that is what I had originally but unfortunately that breaks with the usage of Interruptions to stop this thread. E.g. the API states that it will try and finish any messages that it is currently processing before shutting down. If I have the message being processed in this thread any blocking call in the message implementing method will throw the interrupted exception. The way to get around this was run it in its own thread and let this controller control when it should interrupt the executing thread.

Any suggestions to get around this? I should also document what I explained above.